pokemon-service/src/test/java/cz/muni/fi/pa165/pokemon/league/participation/manager/service/PokemonSpeciesServiceTest.java

```diff
@@ -0,0 +1,185 @@
+        when(pokemonSpeciesDAO.findPokemonSpeciesById(pokemonSpeciesPikachu.getId())).thenReturn(pokemonSpeciesPikachu);
+        when(pokemonSpeciesDAO.findPokemonSpeciesById(pokemonSpeciesRaichu.getId())).thenReturn(pokemonSpeciesRaichu);
+        when(pokemonSpeciesDAO.findPokemonSpeciesById(exceptionalPokemonSpecies.getId())).thenThrow(PE);
+        when(pokemonSpeciesDAO.getAllEvolutionsOfPokemonSpecies(pokemonSpeciesPichu)).thenReturn(Stream.of(pokemonSpeciesPikachu).collect(Collectors.toList()));
+        when(pokemonSpeciesDAO.getAllEvolutionsOfPokemonSpecies(pokemonSpeciesPikachu)).thenReturn(Stream.of(pokemonSpeciesRaichu).collect(Collectors.toList()));
+        when(pokemonSpeciesDAO.getAllEvolutionsOfPokemonSpecies(pokemonSpeciesRaichu)).thenReturn(new ArrayList<>());
+        when(pokemonSpeciesDAO.getAllEvolutionsOfPokemonSpecies(exceptionalPokemonSpecies)).thenThrow(PE);
+        when(pokemonSpeciesDAO.getAllPokemonSpecies()).thenReturn(Stream.of(pokemonSpeciesPichu, pokemonSpeciesRaichu, pokemonSpeciesPikachu).collect(Collectors.toList()));
+        doThrow(PE).when(pokemonSpeciesDAO).updatePokemonSpecies(exceptionalPokemonSpecies);
+        doThrow(PE).when(pokemonSpeciesDAO).deletePokemonSpecies(exceptionalPokemonSpecies);
+    }
+
+    @Test
+    public void createPokemonSpeciesTest() throws EvolutionChainTooLongException {
+        doThrow(PE).when(pokemonSpeciesDAO).createPokemonSpecies(exceptionalPokemonSpecies);
+        assertThatExceptionOfType(PersistenceException.class)
+                .isThrownBy(() -> pokemonSpeciesService.createPokemonSpecies(exceptionalPokemonSpecies));
+        assertThatExceptionOfType(EvolutionChainTooLongException.class)
+                .isThrownBy(() -> pokemonSpeciesService.createPokemonSpecies(pokemonSpeciesPreevolutionTwo));
+
+        pokemonSpeciesService.createPokemonSpecies(pokemonSpeciesPichu);
+        verify(pokemonSpeciesDAO, atLeastOnce()).createPokemonSpecies(pokemonSpeciesPichu);
+    }
+
+    @Test
+    public void changeTypingTest() {
+        assertThatExceptionOfType(PersistenceException.class)
+                .isThrownBy(() -> pokemonSpeciesService.changeTyping(exceptionalPokemonSpecies, PokemonType.BUG, PokemonType.DARK));
+        pokemonSpeciesService.changeTyping(pokemonSpeciesPichu, PokemonType.DARK, PokemonType.DRAGON);
+        verify(pokemonSpeciesDAO, atLeastOnce()).updatePokemonSpecies(pokemonSpeciesPichu);
+        assertThat(pokemonSpeciesPichu.getPrimaryType()).isEqualTo(PokemonType.DARK);
+        assertThat(pokemonSpeciesPichu.getSecondaryType()).isEqualTo(PokemonType.DRAGON);
+    }
+
+    @Test
+    public void changePreevolutionTest() throws EvolutionChainTooLongException, CircularEvolutionChainException {
+        assertThatExceptionOfType(EvolutionChainTooLongException.class)
+                .isThrownBy(() -> pokemonSpeciesService.changePreevolution(pokemonSpeciesPichu, pokemonSpeciesPreevolutionTwo));
```

Comment:
The way your "preevolution" species stack, it is equally right to throw EvolutionChainTooLongException as well as CircularEvolutionChainException, since it is both a circle, and it has more than 3 members - change it so that only one condition is satisfied.